### DISCUSSION TITLE
List of deprecations

### DISCUSSION THREAD

**Discussion by samdark** [yiisoft/yii2#19876] [Yii 22]
There are some places in Yii2 that were marked as deprecated but were never removed since we did no other major version: https://github.com/search?q=repo%3Ayiisoft%2Fyii2%20deprecated&type=code

If you know/remember more, please post here.

**Comment by tomaszkane** [comment]
Maybe some of this:

- https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/db/SchemaBuilderTrait.php#L297
- https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/db/Command.php#L373
- https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/UPGRADE.md?plain=1#L510
- https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/base/View.php#L418

**Comment by smirnov-e** [comment]
I hope you will not delete ability [ability](https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/db/pgsql/ColumnSchema.php#L51)  to turn off parsing arrays to ArrayExpression as default value was quite uncomfortable to use.

**Reply by mtangoo** [reply]
Can you post the  method/line of code  you are referring to?

**Reply by smirnov-e** [reply]
https://github.com/yiisoft/yii2/blob/b073daa8f0e9ea1260cb2f609ba0b811a16f672d/framework/db/pgsql/ColumnSchema.php#L104

**Reply by mtangoo** [reply]
I see. Why is that option uncomfort to you and why you think alternative is the best? Can you make the case for keeping it? 
I think from there we can discuss to conclusion. Reading the docs it seems like it was only for easing migration

**Reply by smirnov-e** [reply]
Well, as I remember from times of migrating 2.0.13 -> 2.0.2x: it was a bad imitation of array, so adding values and usage of `array_*` functions was far from convenient.
Though we sometimes used ArrayExpression to create `@>` criteria when column type was `smallint[]` instead of `int[]` but for all other cases there were no drawbacks in `legacy` way and no profit for `modern`. Maybe this is the only case when default behavior [possibly] performs better than `legacy` one.

**Reply by mtangoo** [reply]
I see. But if you need arrays you can use `getValue()` or did I miss something?
https://www.yiiframework.com/doc/api/2.0/yii-db-arrayexpression#getValue()-detail

**Reply by smirnov-e** [reply]
Yes, you can, but most likely you want to save it back so you will have to recreate ArrayExpression each time.

Also, I may be mistaken, but  as remember operations like `$model->array_attribute[] = $newScalarValue` and `unset($model->array_attribute[1])` also required juggling with getValue + new ArrayExpression. But this part needs verification as it was too long ago.

**Reply by mtangoo** [reply]
Why recreate? you can use `offsetSet` to set a changes array. Or did I miss something?

**Reply by smirnov-e** [reply]
Well, you can.
I've checked some use cases:

1. Accepting new data via POST/GET will replace ArrayExpression with plain array. So you have to make your validation tolerant to such changes / change type on `->load`.
2. It doesn't support multidimensional arrays `$model->array = new ArrayExpression([]); $model->array[1][2] = 'new data'` throws exception.
And let's return to the start: what is the profit of ArrayExpression over plain arrays? I don't know any and that should be enough to drop conversion to ArrayExpression by default.

**Reply by mtangoo** [reply]
>what is the profit of ArrayExpression over plain arrays

Well @samdark can comment on that as I have never needed to call that method directly. TBH I don't even see why you would need to use that method if you use plain arrays. When I use array field, I just serialize it to JSON text and back. So perhaps am not the most qualified to comment on this question

**Reply by smirnov-e** [reply]
Well, when you use JSON columns by default you get JSON (array or scalar), not JSONExpression, no problem there )
And I happy that since 2.0.14 / 2.0.15 you don't have to convert array/JSON column values from string to appropriate type manually.

**Comment by samdark** [comment]
Another one is CUBRID support: https://github.com/yiisoft/yii2/issues/19890

**Reply by jiaweipan** [reply]
Methods mset, mget and madd of yii\caching\Cache, getHasChanged of yii\caching\Dependency: #19892
